src/librustc_resolve/lib.rs

```diff
@@ -3457,6 +3536,123 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+    /// the supplied predicate. The lookup spans across all crates. This method
+    /// should be used to generate suggestions when name resolution fails.
+    ///
+    /// NOTE: We are not looking into imports, so we will ignore all private-
+    /// made-public entities through patterns like this:
+    /// ```
+    /// pub mod foo {
+    ///     mod bar { trait T {}; };
+    ///     // make private T accessible through an import:
+    ///     pub use bar::T as T;
+    /// }
+    fn lookup_candidates<FilterFn>(&mut self,
+                                   lookup_name: Name,
+                                   namespace: Namespace,
+                                   filter_fn: FilterFn) -> SuggestedCandidates
+        where FilterFn: Fn(Def) -> bool {
+
+        let mut lookup_results_accessible = Vec::new();
+        let mut lookup_results_everything = Vec::new();
+        let mut lookup_modules = Vec::new();
+        lookup_modules.push((self.graph_root, Vec::new(), true));
+
+        while let Some((in_module, path_segments, module_is_public)) = lookup_modules.pop() {
+            build_reduced_graph::populate_module_if_necessary(self, &in_module);
+
+            in_module.for_each_child(|name, ns, name_binding| {
+
+                // avoid imports entirely
+                if name_binding.is_import() { return; }
+
+                // collect results based on the filter function
+                if let Some(def) = name_binding.def() {
+                    if name == lookup_name && ns == namespace && filter_fn(def) {
+                        // create the path
+                        let ident = hir::Ident::from_name(name);
+                        let params = PathParameters::none();
+                        let segment = PathSegment {
+                            identifier: ident,
+                            parameters: params,
+                        };
+                        let span = name_binding.span.unwrap_or(syntax::codemap::DUMMY_SP);
+                        let mut segms = path_segments.clone();
+                        segms.push(segment);
+                        let segms = HirVec::from_vec(segms);
+                        let path = Path {
+                            span: span,
+                            global: true,
+                            segments: segms,
+                        };
+                        // push result
+                        // the entity is not accessible in the following cases:
+                        // 1. if it's defined in the same crate, it's always
+                        // accessible
+                        // 2. if it's defined in another crate, it's accessible
+                        // only if both the module is public and the entity is
+                        // declared as public
+                        lookup_results_everything.push(path.clone());
+                        if in_module.is_successor_of(self.current_module) ||
```

Comment:
If we show everything in the current crate, there's no need for `is_descendent_of`.